logging/structured/src/main/scala/tofu/logging/LoggingMid.scala

```diff
@@ -0,0 +1,157 @@
+package tofu.logging
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+import tofu.syntax.monadic._
+import tofu.higherKind.Mid
+import cats.Monad
+import tofu.higherKind.derived.HigherKindedMacros
+import tofu.Errors
+import tofu.syntax.handle._
+
+/** Logging middleware
+  * Alg[LoggingMid] is a special form of implicit evidence of injectable logging support
+  * generally you don't need `Logging` instance to derive this
+  * so choice of logging postponed until this middleware is attached to the core instance
+  */
+abstract class LoggingMid[A] {
+  def around[F[_]: Monad: LoggingBase](fa: F[A]): F[A]
+
+  def toMid[F[_]: Monad: LoggingBase]: Mid[F, A] = around(_)
+}
+
+object LoggingMid extends LoggingMidBuilder.DefaultImpl {
+  type Result[A] = LoggingMid[A]
+  def instance[U[_[_]]]: U[LoggingMid] = macro HigherKindedMacros.factorizeThis[U]
+
+  type Of[U[_[_]]] = U[LoggingMid]
+}
+
+/** Logging middleware generator */
+trait LoggingMidBuilder {
+
+  /** do some logging upon enter to method invocation */
+  def onEnter[F[_]: LoggingBase](cls: Class[_], method: String, args: Seq[(String, LoggedValue)]): F[Unit]
+
+  /** do some logging after leaving method invocation with known result */
+  def onLeave[F[_]: LoggingBase](
+      cls: Class[_],
+      method: String,
+      args: Seq[(String, LoggedValue)],
+      res: LoggedValue
+  ): F[Unit]
+
+  def prepare[Alg[_[_]]](implicit Alg: ClassTag[Alg[Any]]): Prepared[Alg] = new Prepared[Alg](Alg.runtimeClass)
+
+  class Method[U[f[_]], Res: Loggable](
+      cls: Class[_],
+      method: String,
+      args: mutable.Buffer[(String, LoggedValue)]
+  ) {
+    def arg[A: Loggable](name: String, a: A): this.type = {
+      args += (name -> a)
```

Comment:
Shouldn't this code be marked as unsafe somehow with comments or something? 
The reader would need to understand the intentions behind the decision to make it mutable and such.

Reply:
I suppose we can split this unsafe behaviour into the package private implementation of the abstract class, as OOP grandfather behest says